runtime-sdk/modules/evm/src/precompile/subcall.rs

```diff
@@ -0,0 +1,522 @@
+use ethabi::{ParamType, Token};
+use evm::{
+    executor::stack::{PrecompileFailure, PrecompileHandle, PrecompileOutput},
+    ExitError, ExitSucceed,
+};
+
+use crate::backend::EVMBackendExt;
+use oasis_runtime_sdk::{
+    module::CallResult, modules::core::Error, subcall, types::transaction::CallerAddress,
+};
+
+use super::{record_linear_cost, PrecompileResult};
+
+/// A subcall validator which prevents any subcalls from re-entering the EVM module.
+struct ForbidReentrancy;
+
+impl subcall::Validator for ForbidReentrancy {
+    fn validate(&self, info: &subcall::SubcallInfo) -> Result<(), Error> {
+        if info.method.starts_with("evm.") {
+            return Err(Error::Forbidden);
+        }
+        Ok(())
+    }
+}
+
+// XXX: fix these costs
+const SUBCALL_BASE_COST: u64 = 1_000;
+const SUBCALL_WORD_COST: u64 = 1;
+
+pub(super) fn call_subcall<B: EVMBackendExt>(
+    handle: &mut impl PrecompileHandle,
+    backend: &B,
+) -> PrecompileResult {
+    record_linear_cost(
+        handle,
+        handle.input().len() as u64,
+        SUBCALL_BASE_COST,
+        SUBCALL_WORD_COST,
+    )?;
+
+    // Ensure that the precompile is called using a regular call so the caller is actually the
+    // address of the calling contract.
+    if handle.context().address != handle.code_address() {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other("invalid call".into()),
+        });
+    }
+
+    let mut call_args = ethabi::decode(
+        &[
+            ParamType::Bytes, // method
+            ParamType::Bytes, // body (CBOR)
+        ],
+        handle.input(),
+    )
+    .map_err(|e| PrecompileFailure::Error {
+        exit_status: ExitError::Other(e.to_string().into()),
+    })?;
+
+    // Parse raw arguments.
+    let body = call_args.pop().unwrap().into_bytes().unwrap();
+    let method = call_args.pop().unwrap().into_bytes().unwrap();
+
+    // Parse body as CBOR.
+    let body = cbor::from_slice(&body).map_err(|_| PrecompileFailure::Error {
+        exit_status: ExitError::Other("body is malformed".into()),
+    })?;
+
+    // Parse method.
+    let method = String::from_utf8(method).map_err(|_| PrecompileFailure::Error {
+        exit_status: ExitError::Other("method is malformed".into()),
+    })?;
+
+    // Cap maximum amount of gas that can be used.
+    let max_gas = handle.remaining_gas();
+
+    let result = backend
+        .subcall(
+            subcall::SubcallInfo {
+                caller: CallerAddress::EthAddress(handle.context().caller.into()),
+                method,
+                body,
+                max_depth: 8,
```

Comment:
Was this arbitrarily chosen, or is there a reason for depth to be limited at 8? (although i can't imagine needing more or even 8 in a realistic scenario anyway).

Reply:
This was chosen to match the current default for the `contracts` module (where it is actually a parameter that defaults to 8).